lib/UserInterface/ClingCodeCompleteConsumer.h

```diff
@@ -0,0 +1,38 @@
+#ifndef CLINGCODECOMPLETECONSUMER_H
+#define CLINGCODECOMPLETECONSUMER_H
+
+#include "clang/Sema/CodeCompleteConsumer.h"
+#include "clang/Sema/CodeCompleteOptions.h"
+
+#include "textinput/Text.h"
+#include "textinput/Editor.h"
+#include "textinput/Callbacks.h"
+
+using namespace clang;
+using namespace textinput;
+
+class ClingCodeCompleteConsumer : public clang::CodeCompleteConsumer {
+  clang::CodeCompletionTUInfo CCTUInfo;
+public:
+  ClingCodeCompleteConsumer(const CodeCompleteOptions &Opts)
+    : CodeCompleteConsumer(Opts, false), CCTUInfo(new GlobalCodeCompletionAllocator) { }
+  ~ClingCodeCompleteConsumer() {}
+  CodeCompletionAllocator &getAllocator() override { return CCTUInfo.getAllocator();}
+  CodeCompletionTUInfo &getCodeCompletionTUInfo() override { return CCTUInfo; }
+};
+
+class ClingTabCompletion : public textinput::TabCompletion {
+  ClingCodeCompleteConsumer* fCodeCompleteConsumer;
```

Comment:
For better or worse, one year from now, I will not remember what year this code was written :)  ... and hammering the point does not hurt ... so please :)